google/cloud/pubsublite/internal/partition_publisher.h

```diff
@@ -0,0 +1,322 @@
+  void Rebatch(bool process_in_flight) {
+    std::lock_guard<std::mutex> g{mu_};
+    deque<MessageWithFuture> unbatched_messages_with_futures;
+    if (process_in_flight) {
+      FlattenAndMoveBatch(in_flight_batches_, unbatched_messages_with_futures);
+    }
+    FlattenAndMoveBatch(batches_, unbatched_messages_with_futures);
+
+    for (auto& message_with_future : unbatched_messages_with_futures_) {
+      unbatched_messages_with_futures.emplace_back(
+          MessageWithFuture{std::move(message_with_future.message),
+                            std::move(message_with_future.message_promise)});
+    }
+    unbatched_messages_with_futures_.clear();
+
+    // rebatch
+    Batch current_batch;
+    size_t current_byte_size = 0;
+    size_t current_messages = 0;
+    for (auto& message_with_future : unbatched_messages_with_futures) {
+      size_t message_size = message_with_future.message.ByteSizeLong();
+      if (current_messages + 1 > batching_settings_.GetMessageLimit() ||
+          current_byte_size + message_size >
+              batching_settings_.GetByteLimit()) {
+        if (!current_batch.empty()) {
+          batches_.push_back(std::move(current_batch));
+          // clear because current_batch left in 'unspecified state' after move
+          current_batch.clear();
+          current_byte_size = 0;
+          current_messages = 0;
+        }
+      }
+      current_batch.push_back(std::move(message_with_future));
+      current_byte_size += message_size;
+      ++current_messages;
+    }
+    if (!current_batch.empty()) batches_.push_back(std::move(current_batch));
+  }
+
+  void WriteBatches() {
+    promise<void> root;
+    future<void> root_future = root.get_future();
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (batches_.empty() || shutdown_) {
+        writing_ = false;
+        return;
+      }
+      if (!writing_) {
+        writing_ = true;
+      }
+      auto batch = std::move(batches_.front());
+      batches_.pop_front();
+      MessagePublishRequest publish_request;
+      for (auto& message_with_future : batch) {
+        auto* message = publish_request.add_messages();
+        *message = message_with_future.message;
+      }
+      in_flight_batches_.push_back(std::move(batch));
+      root_future
+          .then(ChainFuture(resumable_stream_->Write(std::move(publish_request),
+                                                     grpc::WriteOptions())))
+          .then([this](future<bool> write_response) {
+            if (write_response.get()) WriteBatches();
+            std::lock_guard<std::mutex> g{mu_};
+            writing_ = false;
+          });
+    }
+    root.set_value();
+  }
+
+  void Read() {
+    promise<void> root;
+    future<void> root_future = root.get_future();
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      if (shutdown_) return;
+      root_future.then(ChainFuture(resumable_stream_->Read()))
+          .then([this](future<absl::optional<MessagePublishResponse>>
```

Comment:
This is not the type of the response. It's a PublishResponse, which should always contain a MessagePublishResponse